README
PREREQUISITES
bashhate		(optionnal) automated style checker for bash scripts
bc				arbitrary precision numeric processing language
csvtool			to manipulate CSV files
curl			to upload to FTP server
ffmpeg			for video conversion and drawing text overlays
FitCSVTool.jar	Java archive provided as part of Garmin Fit SDK
				it can be downloaded here https://developer.garmin.com/fit/download/
java			jre version 1.8.0 or higher, to run the file FitCSVTool.jar
jq				lightweight and flexible command-line JSON processor
				used for parsing ffmpeg json outputs
shellcheck 		(optionnal) analysis tool that automatically finds bugs in your shell scripts

HOW TO VERIFY CODE
bashate *.sh \
&& shellcheck --check-sourced --external-sources main.sh \
&& bash -o noexec main.sh

HOW TO RUN
./main.sh


CHANGELOG
2021-07-12
	parcours des fichiers MP4 et GPX
	check si un fichier gpx correspond à une vidéo (dates)
2021-07-13
	_gps_logs_exist
	_upload_clip
2021-07-16
	sampled the expected behavior with 6 metrics (see image)
	split function __is_gps_log_within_clip into multiple functions
	distance POC

2021-07-24
        added fit to csv conversion
	deprecated gps distance calculation
2021-08-14
	debugging infinite loop
2021-20-08
	split _get_files_with_date to use fit names and gopro ffprobe
2021-08-25
	added function __get_files_modified_asc to sort clips
	managed upside-down videos with an empty drawtext in __convert_csv_to_cmd
	moved repetitive call to _get_fit_files_with_date out of the deepest for loop: improved speed
2021-08-26
	added Elapsed time + offset between clips


main.sh
#!/usr/bin/env bash
source constants.sh
source logging.sh
source dates.sh
source video.sh
source gps.sh
source misc.sh

#exits the script when a command fails
set -o errexit
#add || true to commands that you allow to fail.

# exit the script when using undeclared variables.
set -o nounset

trap exit_trap 0 SIGHUP SIGINT SIGQUIT SIGABRT SIGTERM
trap err_trap ERR

#Iterates through all days found in GoPro folder.
# for each clip, add GPS data (text overlay) if Garmin GPS devise is mounted.
# then create a concatenated output_video and upload it to a remote server.
function main {
#set -x
    _set_xtrace_ps4
    _verify_dependencies

    for day in $(_list_days_files_in_dir "${GOPRO_CLIPS_FOLDER}"); do
        __msg_debug "iteration of day ${day}"
        local output_video
        output_video="${WORKING_FOLDER}/${day}.mp4"
        declare -a gps_logs_fit
        gps_logs_fit=()
        for log in $(_get_fit_files_with_date \
                "${GPS_LOGS_FOLDER}" "${day}"); do
            __msg_debug "log=${log}"
            gps_logs_fit+=("${log}")
        done
        #TODO fixme replace __fit_to_csv_records with _fit_to_csv_records_day
        _fit_to_csv_records_day "${day}" "${gps_logs_fit[@]}"
        declare -a clips_to_concatenate
        clips_to_concatenate=()
        local elapsed_offset
        elapsed_offset=0
        for clip in $(_get_clip_files_with_date \
                "${GOPRO_CLIPS_FOLDER}" "${day}"); do
            __msg_debug "clip=${clip}"
            if _is_garmin_mounted ; then
                local ffmpeg_cmd
                ffmpeg_cmd=$(_convert_fit_to_drawtext_cmd \
                    "${clip}"   \
                    "${elapsed_offset}" \
                    "${gps_logs_fit[@]}")

                __msg_debug "ffmpeg_cmd=${ffmpeg_cmd}"
                if [[ -s "${ffmpeg_cmd}" ]]; then
                    #override clip with the overlay version:
                    clip=$(_sendcmd_to_overlay "${clip}" "${ffmpeg_cmd}")
                fi
            fi
            clips_to_concatenate+=("${clip}")
            __msg_debug "clips_to_concatenate=${clips_to_concatenate[*]}"
            elapsed_offset=$((
                elapsed_offset
                +
                $(__get_clip_duration_seconds "${clip}")
                ))
        done
        _concatenate_clips "${output_video}" "${clips_to_concatenate[@]}"
        rm -f "${clips_to_concatenate[@]}"
        _upload_clip "${output_video}"
        rm -f "${output_video}"

    done
}
main


misc.sh
#!/usr/bin/env bash

# displays the sorted-list of days (time of last modification)
#of files in a directory
function _list_days_files_in_dir {
    local directory
    directory="$1"
    stat --format=%y "${directory}"/* | cut -d ' ' -f1 | sort | uniq
}


#Converts a decimal number into a base36 number (alphabet 0-9A-Z)
#$1 decimal number to convert
#Output on STDOUT its base36 equivalent
function __decimal_to_base36 {
    local base36
    #shellcheck disable=SC2207
    base36=($(echo {0..9} {A..Z}))
    for i in $(bc <<< "obase=36; $1"); do
        echo -n "${base36[$(( 10#$i ))]}"
    done && echo
}


#displays fit filenames with names that match date
# follows this logic:
# https://zerobin.net/?b2bfb1be32a4dc12#5bHF9ub81tJK1bDCifGJJvGHjd2uM0/A/efXEouWzyo=
# $1 directory of fit files
# $2 the date with format yyyy-MM-dd
function _get_fit_files_with_date {
    local directory
    directory="$1"
    local date
    date="$2"
    local char1_year
    char1_year="${date:2:2}"
    char1_year=$((char1_year-10))
    char1_year=$(__decimal_to_base36 "${char1_year}")
    local char2_month
    char2_month=$(__decimal_to_base36 "${date:5:2}")
    local char3_daynr
    char3_daynr=$(__decimal_to_base36 "${date:8:2}")
    local pattern
    pattern="${char1_year}${char2_month}${char3_daynr}*"
    find "${directory}" -name "${pattern}"
}

#Displays the list of files ordered by modification_date ascending
# $1 the directory in which files are located
#Output STDOUT one line per file path
function __get_files_modified_asc {
    local directory
    directory="$1"
    #option --format=%y,%n generates a CSV with modification date,filename (with dir)
    stat --format='%y,%n' "${directory}"/*   \
    | sort                               \
    | cut -d ',' -f2 #suppress dates from the output, only show filenames
}



#Displays clip filenames with dates that match the input date
# $1 directory of clip files
# $2 the date with format yyyy-MM-dd
function _get_clip_files_with_date {
    local directory
    directory="$1"
    local date
    date="$2"

    local ret
    ret=""

    for clip in $(__get_files_modified_asc "$directory"); do
        local start_day
        start_day=$(__get_clip_start_date_iso "${clip}")

        if [ "${start_day}" = "${date}" ]; then
            ret+="${clip} "
        fi
    done
    echo "${ret}"
}


function _verify_dependencies {
    for dependency in ${DEPENDENCIES}; do
    if ! command -v "${dependency}" &> /dev/null
    then
        __msg_error "dependency ${dependency} could not be found"; exit 1
    fi
    done
}



function _is_garmin_mounted {
    if ! [[ -d ${GPS_LOGS_FOLDER} ]]; then
    __msg_debug "unable to find garmin directory"
    return 1
    else
    return 0
    fi
}

#Removes BOM character from a UTF-8 file
# BOM means Byte order mark
# see doc here https://en.wikipedia.org/wiki/Byte_order_mark
#
#$1 the input file from which BOM must be removed
#Output to STDOUT the content of the file, BOM removed
function ___remove_BOM {
    sed -i '1s/^\xEF\xBB\xBF//' "$1"
}

#Removes the 1st line of a file and write it to a new file named $1_noheader
#Globals:
#Arguments: $1 the input file
#Outputs:
#Returns: the return value of tail
function __remove_header {
    tail -n +2 "$1" >| "${1}_noheader"
}

#Traps any exit signals SIGHUP SIGINT SIGQUIT SIGABRT SIGTERM
#and cleans up temporary files
# SIGHUP (“hang-up”) is used to report that the user’s terminal is disconnected,
# SIGINT (“program interrupt”) is sent when the user types the INTR character (normally C-c)
# SIGQUIT is similar to SIGINT, except that it’s controlled by a different key
#  the QUIT character, usually C-\—
#  and produces a core dump when it terminates the process
# SIGABRT is commonly used by libc and other libraries to abort the program in case of
# critical errors.
# SIGTERM is a generic signal used to cause program termination.
#  Unlike SIGKILL, this signal can be blocked, handled, and ignored.
#  It is the normal way to politely ask a program to terminate.
#  The shell command kill generates SIGTERM by default.
#Globals:
#Arguments:
#Outputs:
#Returns:
function exit_trap {
#TODO uncomment exit_trap rm -rf "${WORKING_FOLDER}"
#    rm -rf "${WORKING_FOLDER}"
    exit
}

#Traps ERR fake signal (A command returning a non-zero exit status)
# and display the command name and exit status
#Globals:
#Arguments:
#Outputs: on STDERR the command name and exit code
#Returns true (echo return code)
function err_trap {
    #TODO test err_trap
    local exit_status="$?"
    echo "ERROR: Command exited with status ${exit_status}."
    exit "${exit_status}"
}


video.sh
#!/usr/bin/env bash

#Adds overlay text to a video file, using ffmpeg sendcmd filter
#see documentation: http://ffmpeg.org/ffmpeg-filters.html#sendcmd_002c-asendcmd
#$1 the clip file path
#$2 the ffmpeg cmd file path
#Outputs to STDOUT the new clip file path
function _sendcmd_to_overlay {
    local clip
    clip="$1"
    local ffmpeg_cmd
    ffmpeg_cmd="$2"

    if [[ -s "${ffmpeg_cmd}" ]]; then
    local output_video
    output_video=$(basename "${clip}")
    output_video="${WORKING_FOLDER}/${output_video}_with_overlay.mp4"
        local ffmpeg_command
    ffmpeg_command="ffmpeg ${FFMPEG_OPTS} \
    ${FFMPEG_DEBUG_OPTS} \
    -i \"${clip}\"       \
    -vf \"sendcmd=f=${ffmpeg_cmd},drawtext=text=:${DRAWTEXT_FORMAT}\" \
    \"${output_video}\""
    __msg_debug "ffmpeg_command=${ffmpeg_command}"
    eval "${ffmpeg_command}"
    clip="${output_video}"
    fi
    echo "${clip}"
}

#Concatenates multiple video clips into one video clip
# $1 the output video path
# $@ clips to concatenate (array)
#Outputs to STDOUT/STDERR ffmpeg various outputs
function _concatenate_clips {
    local output
    output="$1"
    __msg_debug "output=${output}"
    shift

    local file_list_inputs
    file_list_inputs=$(mktemp  -p "${WORKING_FOLDER}" \
        -t "clips_to_concat-$(date +%Y-%m-%d_%H.%M.%S).XXX")

    #TODO shorten array writing with printf
    # printf "file '%s'\n" "$@" > "${file_list_inputs}"
    # https://trac.ffmpeg.org/wiki/Concatenate
    for clip in "$@"; do
        __msg_debug "clip : ${clip}"
        echo "file '${clip}'" >> "${file_list_inputs}"
    done

    local ffmpeg_command
    ffmpeg_command="ffmpeg  \
            ${FFMPEG_OPTS}           \
            -f concat -safe 0            \
        -i ${file_list_inputs}       \
        -c copy ${output}"

    __msg_debug "ffmpeg_command=${ffmpeg_command}"
    eval "${ffmpeg_command}"
}



# shows the format of a clip, formatted in json
# $1 the clip path
function ___get_clip_format {
    ffprobe -loglevel error     \
        -print_format json  \
        -show_format    \
        "$1"
}


# shows the duration of a clip in seconds
# $1 the clip path
function __get_clip_duration_seconds {
    local ret
    ret=$(___get_clip_format "$1" | jq -r '.format.duration')
    # rounds duration down to the second (quotient)
    # I use bc instead of bash arithmetic operation because it's a float number
    ret=$(bc <<< "${ret} / 1")
    echo "${ret}"
}

# Shows the start date of a clip in format yyyy-MM-dd
# $1 the clip path
function __get_clip_start_date_iso {
    local ret
    ret=$(___get_clip_format "$1" | jq -r '.format.tags.creation_time')
    echo "${ret:0:10}"
}

# Shows the start date of a clip in seconds
# $1 the clip path
function __get_clip_start_date {
    local ret
    ret=$(___get_clip_format "$1" | jq -r '.format.tags.creation_time')

    ret=$(___date_to_seconds "$ret")

    ret=$((ret + GOPRO_DATE_ADJUSTMENT_SEC))
    echo "${ret}"
}

# Shows the end date of a clip
# $1 the clip path
function __get_clip_end_date {
    local clip_start_date
    clip_start_date=$(__get_clip_start_date "$1")

    local clip_duration
    clip_duration=$(__get_clip_duration_seconds "$1")

    local clip_end_date
    clip_end_date=$((clip_start_date + clip_duration))
    echo "${clip_end_date}"
}

# Uploads a video clip to FTP server
# $1 - local path of the clip to upload
function _upload_clip {
    local clip_path
    clip_path="$1"
    local ftp_password
    ftp_password=$(base64 -d <<< "${FTP_PASSWORD_BASE64}")
    curl --silent --user "${FTP_USERNAME}":"${ftp_password}" \
         --upload-file "${clip_path}" ftp://"${FTP_HOSTNAME}"
}


logging.sh
#!/usr/bin/env bash

#Description: prints a message, only if constant ERROR is set to 1
#Globals:
#Arguments: $* the strings to print
#Outputs: the message is shown to STDERR
#Returns:
function __msg_error {
    [[ "${ERROR}" == "1" ]] && echo -e "$(date '+%F %T.%3N') [ERROR] $*" 1>&2
}

#Description: prints a message, only if constant DEBUG is set to 1
#Globals:
#Arguments: $* the strings to print
#Outputs: the message is shown to STDERR
#Returns:
function __msg_debug {
    [[ "${DEBUG}" == "1" ]] && echo -e "$(date '+%F %T.%3N') [DEBUG] $*" 1>&2
}

#Description: prints message, only if constant WARN is set to 1
#Globals:
#Arguments: $* the strings to print
#Outputs: the message is shown to STDERR
#Returns:
function __msg_warn {
    [[ "${DEBUG}" == "1" ]] && echo -e "$(date '+%F %T.%3N') [WARN] $*" 1>&2
}

#Sets a nicer output for Bash set -o xtrace (a.k.a set -x)
#doc here https://wiki.bash-hackers.org/scripting/debuggingtips
# : information separator
# + the character repeated for multiple levels of expansion
# ${BASH_SOURCE[0]} the name of current script file file being sourced
# ${FUNCNAME[0]} the current function name
# ${LINENO} the line number
#Globals:
#Arguments:
#Outputs: $PS4 is written to STDOUT if xtrace mode is enabled
#Returns:
function _set_xtrace_ps4 {
    #PS4="+$(date '+%F %T.%3N'):${BASH_SOURCE[0]}:${LINENO}:${FUNCNAME[0]} "
    export PS4='+(${BASH_SOURCE}:${LINENO}):${FUNCNAME[0]:+${FUNCNAME[0]}():}'
}


constants.sh
#!/usr/bin/env bash
export DEPENDENCIES="java jq ffmpeg bc curl csvtool"
export ERROR="1"
export INFO="1"
export DEBUG="1"

if [[ "${DEBUG}" == "1" ]] ; then
    FFMPEG_OPTS="-hide_banner"
else
    FFMPEG_OPTS="-hide_banner -loglevel error"
fi
export FFMPEG_OPTS
#reduces time of output videos to 1s, encode to aprox. 20sec
#export FFMPEG_DEBUG_OPTS="-t 10"
export FFMPEG_DEBUG_OPTS=""

WORKING_FOLDER=$(mktemp -d -p "/tmp/" \
    -t "gopro-$(date +%Y-%m-%d_%H.%M.%S)-XXXXXXXXXX")
export WORKING_FOLDER

#export GOPRO_CLIPS_FOLDER="$HOME/gopro/Samples/2021-07-31_indoors3/GOPRO/"
GOPRO_PATH="/run/user/1000/gvfs/mtp:host=GoPro_HERO7_C3281328000793/"
GOPRO_PATH+="GoPro MTP Client Disk Volume/DCIM/100GOPRO/"
ln -s "${GOPRO_PATH}" "${WORKING_FOLDER}/link_to_GoPro"
GOPRO_CLIPS_FOLDER="${WORKING_FOLDER}/link_to_GoPro"
#GOPRO_CLIPS_FOLDER="$HOME/gopro/Samples/GoPro/2021-08-15.1/"
export GOPRO_CLIPS_FOLDER
export GARMIN_FITCSVTOOL_JAR="$HOME/gopro/Garmin_FitSDK/java/FitCSVTool.jar"
#export GPS_LOGS_FOLDER="$HOME/gopro/Samples/Garmin/"
export GPS_LOGS_FOLDER="/media/pi/GARMIN/GARMIN/ACTIVITY/"
export GPS_LOGS_NAME_PATTERN="*.FIT"
export GOPRO_DATE_ADJUSTMENT_SEC=$((-2 * 3600))
SECONDS_BETWEEN_UNIX_AND_GARMIN_EPOCH=$(date -d 'UTC 00:00 Dec31 1989' '+%s')
export SECONDS_BETWEEN_UNIX_AND_GARMIN_EPOCH

export DATAFIELD_CADENCE="1"
export DATAFIELD_DISTANCE="2"
export DATAFIELD_GRADE="3"
export DATAFIELD_SPEED="4"
export DATAFIELD_HEARTRATE="5"
export DRAWTEXT_FORMAT='
fontfile=/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf:
fontsize=20:
fontcolor=white:
box=1:
boxcolor=black@0.7:
boxborderw=10:
x=10:
y=main_h-(6*line_h)-10:
'
export FTP_HOSTNAME="vps719805.ovh.net"
export FTP_PORT="21"
export FTP_USERNAME="axis-camera"
export FTP_PASSWORD_BASE64="JGI3QDhlUk4jdlk="


gps.sh
#!/usr/bin/env bash

# converts a FIT log file into a CSV file
# COLNUM COLNAME                  EXAMPLE VALUE           COMMENT
# 01 timestamp[s]                 "995990727"             USE for filtering
# 02 position_lat[semicircles]    "582881112"             IGNORE
# 03 position_long[semicircles]   "27482251"              IGNORE
# 04 distance[m]                  "11.83"                 USE
# 05 altitude[m]                  "182.39999999999998"    USE (grade %)
# 06 speed[m/s]                   "4.031"                 USE
# 07 unknown                                              IGNORE
# 08 heart_rate[bpm]              "102"                   USE
# 09 cadence[rpm]                 "54"                    USE
# 10 temperature[C]               "30"                    IGNORE
# 11 fractional_cadence[rpm]      "0.0"                   IGNORE
# 12 enhanced_altitude[m]         "182.39999999999998"    IGNORE
# 13 enhanced_speed[m/s]          "4.031"                 IGNORE
#
# $1 the input FIT file path
# $2 the output file path without suffix
#Output STDOUT: the output file path, ending with suffix "_data.csv"
function __fit_to_csv {
    local fit_file
    fit_file="$1"
    local csv_file
    csv_file="$2"

    java -jar "${GARMIN_FITCSVTOOL_JAR}" -b "${fit_file}" \
        "${csv_file}" --defn "none" --data "record" 1> "/dev/null"
    #I don't need this "dirty" file automatically generated
    rm "${csv_file}.csv"
    #I do need this one:
    csv_file="${csv_file}_data.csv"
    ___remove_BOM "${csv_file}"
    echo "${csv_file}"
}

#Converts a CSV file into ffmpeg file
# $1 the input concatenated csv file path
# $2 the output ffmpeg command file path
# $3 the elapsed time offset in seconds
function __convert_csv_to_cmd {
    local concatenated_csv_data
    concatenated_csv_data="$1"
    local cmd_file
    cmd_file="$2"
    local elapsed_offset
    elapsed_offset="$3"
    #write at least one line, useful if there is no gps data
    echo "0-0 [leave] drawtext reinit 'text=';" > "${cmd_file}"
    local interval_start
    interval_start=0
    #shellcheck disable=SC2034
    while IFS=, read -r ts dist alt speed cadence hr; do
        printf "%d-%d [enter] drawtext reinit 'text=\\n" \
               "${interval_start}" $((interval_start + 1))

        printf "Elapsed      %02d\\:%02d\\:%02d\\n"                   \
               $(( (interval_start + elapsed_offset) / 3600 ))        \
               $(( ( (interval_start + elapsed_offset) / 60 ) % 60 )) \
               $(( (interval_start + elapsed_offset) % 60 ))
        printf "Cadence      %d RPM\\n" \
               "${cadence}"
        dist=$(bc <<< "scale=2 ; ${dist} / 1000")
        printf "Distance     %0.2f Km\\n"                 \
               "${dist}"

        #TODO add Grade metric (with %unit) based on alt and dist
        printf "Grade         %+.2f (random for now)\\n" \
               $((-10 + RANDOM % 20))
        speed=$(bc <<< "${speed} * 3.6")
        printf "Speed        %0.f KPH\\n" \
               "${speed}"
        printf "Heart rate     %d BPM\\n" \
               "${hr}"
        printf "';\\n"
        interval_start=$((interval_start + 1))

    done < "${concatenated_csv_data}" >> "${cmd_file}"
}


#Converts a list of FIT files into a "concatenated" CSV file
# $1 the day with format yyyy-MM-dd
# $@ all fit files
# outputs to STDOUT the path to clean CSV record file
function _fit_to_csv_records_day {
    local day
    day="$1"
    shift

    #warning: this is an array variable
    declare -a logs_gps_csv_data
    logs_gps_csv_data=()
    #shellcheck disable=SC2068
    for fit_file in $@; do
    local csv_file
    csv_file=$(basename "${fit_file}")
    csv_file="${WORKING_FOLDER}/${csv_file}"

    csv_file=$(__fit_to_csv "${fit_file}" "${csv_file}")

    local namedcol
    namedcol='record.timestamp[s]'
    namedcol+=',record.distance[m]'
    namedcol+=',record.altitude[m]'
    namedcol+=',record.speed[m/s]'
    namedcol+=',record.cadence[rpm]'
    #add column 'record.heart_rate[bpm]' only if available
    if grep --quiet 'record.heart_rate' "${csv_file}" ; then
        namedcol+=',record.heart_rate[bpm]'
    fi


    csvtool namedcol "${namedcol}" "${csv_file}" \
        > "${csv_file}_interesting_cols"

    csv_file="${csv_file}_interesting_cols"
        __remove_header "${csv_file}"
        csv_file="${csv_file}_noheader"
    logs_gps_csv_data+=("${csv_file}")
    done

    local concatenated_csv_data
    concatenated_csv_data="${WORKING_FOLDER}/${day}_gps_data_concat.csv"
    if (( ${#logs_gps_csv_data[@]} > 0)); then
    cat "${logs_gps_csv_data[@]}" > "${concatenated_csv_data}"
    fi
    echo "${concatenated_csv_data}"
}


#converts a list of fit files into a "concatenated" CSV records
# $1 clip_basename the file name of the clip
# $@ all fit files
# outputs to STDOUT the path to clean CSV record file
function __fit_to_csv_records {
    local clip_basename
    clip_basename="$1"
    shift

    #warning: this is an array variable
    declare -a logs_gps_csv_data
    logs_gps_csv_data=()
    #shellcheck disable=SC2068
    for fit_file in $@; do
    local csv_file
    csv_file=$(basename "${fit_file}")
    csv_file="${WORKING_FOLDER}/${csv_file}"

    csv_file=$(__fit_to_csv "${fit_file}" "${csv_file}")

    #csvtool col 1,4,5,6 "${csv_file}" > "${csv_file}_interesting_cols"

        local namedcol
    namedcol='record.timestamp[s]'
    namedcol+=',record.distance[m]'
    namedcol+=',record.altitude[m]'
    namedcol+=',record.speed[m/s]'
    namedcol+=',record.cadence[rpm]'
    #add column 'record.heart_rate[bpm]' only if available
    if grep --quiet 'record.heart_rate' "${csv_file}" ; then
        namedcol+=',record.heart_rate[bpm]'
    fi

    csvtool namedcol "${namedcol}" "${csv_file}" \
        > "${csv_file}_interesting_cols"

    csv_file="${csv_file}_interesting_cols"
        __remove_header "${csv_file}"
        csv_file="${csv_file}_noheader"
    logs_gps_csv_data+=("${csv_file}")
    done

    local concatenated_csv_data
    concatenated_csv_data="${WORKING_FOLDER}/${clip_basename}_concat.csv"
    if (( ${#logs_gps_csv_data[@]} > 0)); then
    cat "${logs_gps_csv_data[@]}" > "${concatenated_csv_data}"
    fi
    echo "${concatenated_csv_data}"
}

#Converts a list of FIT files into short concatenated CSV files
#Globals:
#Arguments:
# $1 clip file (used for date matching and output filename)
# $2 the elapsed time offset in seconds
# $@ all fit files of the same day
#Outputs: the filename that can be used by ffmpeg drawtext filter
#Returns:
function _convert_fit_to_drawtext_cmd {
    local clip
    clip="$1"
    local clip_basename
    clip_basename=$(basename "${clip}")

    local elapsed_offset
    elapsed_offset="$2"
    shift
    shift

    local concatenated_csv_data
    concatenated_csv_data=$(__fit_to_csv_records "${clip_basename}" "$@")
    __msg_debug "concatenated_csv_data=${concatenated_csv_data}"
    local start
    start=$(__get_clip_start_date "${clip}")
    start=$(__unix_to_garmin_ts "${start}")
    local end
    end=$(__get_clip_end_date "${clip}")
    end=$(__unix_to_garmin_ts "${end}")

    #filter log lines having their first column (awk's $1) between $start and $end
    awk "${start} < \$1 && \$1 < ${end}" \
    < "${concatenated_csv_data}" \
    > "${concatenated_csv_data}_filtered"
    __msg_debug "concatenated_csv_data_filtered=
    ${concatenated_csv_data}_filtered}"

    local cmd_file
    cmd_file="${WORKING_FOLDER}/${clip_basename}.cmd_ffmpeg"
    __convert_csv_to_cmd "${concatenated_csv_data}_filtered" \
                         "${cmd_file}" \
                         "${elapsed_offset}"
    echo "${cmd_file}"
}
